File: datasets/C++/Mistral/DataStructures/medium5.cpp
#include <iostream>
#include <vector>
#include <sstream>
#include <climits>
using namespace std;

struct Node {
    int data;
    Node* left;
    Node* right;
    Node(int val) : data(val), left(nullptr), right(nullptr) {}
};

bool isBST(Node* node, int min_val = INT_MIN, int max_val = INT_MAX) {
    if (node == nullptr) {
        return true;
    }
    if (node->data <= min_val || node->data >= max_val) {
        return false;
    }
    return isBST(node->left, min_val, node->data) && isBST(node->right, node->data, max_val);
}

Node* insertLevelOrder(const vector<int>& arr, int i) {
    if (i < arr.size() && arr[i] != -1) {
        Node* root = new Node(arr[i]);
        root->left = insertLevelOrder(arr, 2 * i + 1);
        root->right = insertLevelOrder(arr, 2 * i + 2);
        return root;
    }
    return nullptr;
}

int main() {
    cout << "Enter the tree nodes in level order (use -1 for None), separated by spaces: ";
    string input;
    getline(cin, input);
    istringstream iss(input);
    vector<int> arr;
    int value;
    while (iss >> value) {
        arr.push_back(value);
    }
    Node* root = insertLevelOrder(arr, 0);
    cout << (isBST(root) ? "Yes" : "No") << endl;
    return 0;
}

File: datasets/Python/Mistral/DataStructures/medium5.py
class Node:
    def __init__(self, data):
        self.data = data
        self.left = None
        self.right = None

def is_bst(node, min_val=float('-inf'), max_val=float('inf')):
    if node is None:
        return True
    if not (min_val < node.data < max_val):
        return False
    return is_bst(node.left, min_val, node.data) and is_bst(node.right, node.data, max_val)

def insert_level_order(arr, root, i):
    if i < len(arr) and arr[i] != -1:
        root = Node(arr[i])
        root.left = insert_level_order(arr, root.left, 2 * i + 1)
        root.right = insert_level_order(arr, root.right, 2 * i + 2)
    return root

def main():
    arr = list(map(int, input("Enter the tree nodes in level order (use -1 for None): ").split()))
    root = None
    root = insert_level_order(arr, root, 0)
    print("Yes" if is_bst(root) else "No")

if __name__ == "__main__":
    main()
